### ISSUE TITLE
cuckoohashmap taking high memory 

### ISSUE THREAD

**Issue by naidud2020** [efficient/libcuckoo#136]
Tried with cuckoohashmap and unordered_map. 
for same data, cuckoohashmap taken morethan 64GB on RAM and crashed. unordered_map finished with 1.5GB of data.
I would like to give preference to libcuckoo hash map. Attaching my sample c++ code. Do let me know whether libcuckoo supports with small memory print like unordered_map.

[nested_table.cc.txt](https://github.com/efficient/libcuckoo/files/6976385/nested_table.cc.txt)

code:

/* We demonstrate how to nest hash tables within one another, to store
 * unstructured data, kind of like JSON. There's still the limitation that it's
 * statically typed. */

#include <iostream>
#include <memory>
#include <string>
#include <utility>
#include <unordered_map>

#include <libcuckoo/cuckoohash_map.hh>

void test_with_cuckuoo_hash()
{
  typedef libcuckoo::cuckoohash_map<std::string, std::string> InnerTable;
  typedef libcuckoo::cuckoohash_map<std::string, std::unique_ptr<InnerTable>> OuterTable;

  OuterTable tbl;
  tbl.reserve(5000000); //reserved

  for(int i = 1000000; i < 6000000; ++ i) {
      std::string outer_key{std::to_string(i)};
      std::string inner_key{std::to_string(i)};
      std::string inner_val{std::to_string(i)};

      tbl.insert(outer_key, std::unique_ptr<InnerTable>(new InnerTable));
      tbl.update_fn(outer_key, [&](std::unique_ptr<InnerTable> &innerTbl) {
          innerTbl->reserve(4); //reserved
          innerTbl->insert(inner_key, inner_val);
      });
  }

  std::cout << "count: " << tbl.size() << std::endl;

}

void test_with_unordered_map()
{
  typedef std::unordered_map<std::string, std::string> InnerTable;
  typedef std::unordered_map<std::string, std::unique_ptr<InnerTable>> OuterTable;

  OuterTable tbl;

  for(int i = 1000000; i < 6000000; ++ i) {
      std::string outer_key{std::to_string(i)};
      std::string inner_key{std::to_string(i)};
      std::string inner_val{std::to_string(i)};

      std::unique_ptr<InnerTable> inner_tbl = std::make_unique<InnerTable>();
      inner_tbl->insert({inner_key, inner_val});
      tbl.insert({outer_key, std::move(inner_tbl)});
  }

  std::cout << "count: " << tbl.size() << std::endl;

}

int main() {
  //test_with_cuckuoo_hash(); //CONSUMING MORE THAN 64 GB, STILL COULDNT COMPLETE 
  test_with_unordered_map();  //CONSUMING 1.5 GB 
  return 0;
}

**Comment by manugoyal** [comment]
Hi @naidud2020, it seems the problem here is that the [default constructor](https://github.com/efficient/libcuckoo/blob/93217f8d391718380c508a722ab9acd5e9081233/libcuckoo/cuckoohash_map.hh#L104) for a cuckoohash table will allocate buckets for [2^18 elements ](https://github.com/efficient/libcuckoo/blob/93217f8d391718380c508a722ab9acd5e9081233/libcuckoo/cuckoohash_config.hh#L15). If I change your code to construct the `InnerTable` as `new InnerTable(1)`, the memory usage on my machine for `test_with_cuckuoo_hash` goes down to about 5.5G.

The thing with the cuckoohash table is that very small tables are not particularly memory efficient, because there are several auxiliary data structures such as a mutex array which must be allocated for each table. So if you don't need thread safety for the inner table, it might be a good idea to use `cuckoohash_map` for the `OuterTable` and `unordered_map` for the inner table. On my machine, this results in ~900MB memory usage for your example.

Hope this helps!

**Comment by naidud2020** [comment]
Hi @manugoyal Thanks for your response.  need thread safe for outer and inner table.

During my experiments, I found that there is a good performance with integer as a key for outer and inner tables. 

Is it recommended to use libcuckoo hash map to store "20 Million Keys in outer table and each outer table has ~5 Inner keys".  Will it support?

**Comment by manugoyal** [comment]
Yes I expect that would work fine. Another option for thread-safety on the inner table is to store something like a `pair<std::mutex, std::unordered_map>`. So a coarse-grained level of thread-safety for the inner table, and a fine-grained level for the outer table. Also if you have a really small number of keys in the inner table, it might be faster to use a simple data structure like `std::vector<std::pair<key value>>`, and just linearly search. But probably worth trying out several options.

**Comment by naidud2020** [comment]
I would like to understand the usecases for libcuckoo hash map. can't have many mutexes as per our requirement.

please share the usecases for "high performance" and "low memory footprint". will change according to libcuckoo usecases.
